src/main/java/org/truffleruby/core/encoding/EncodingNodes.java

```diff
@@ -583,6 +583,43 @@ protected RubyEncoding getActualEncodingDummy(Rope rope, RubyEncoding encoding)
 
             return encoding;
         }
+
+        @TruffleBoundary
+        @Specialization(guards = "encoding.jcoding.isDummy()")
+        protected RubyEncoding getActualEncodingDummy(AbstractTruffleString string, RubyEncoding encoding,
+                @Cached TruffleString.GetInternalByteArrayNode byteArrayNode) {
+            if (encoding.jcoding instanceof UnicodeEncoding) {
+                var enc = encoding.tencoding;
+                var byteArray = byteArrayNode.execute(string, enc);
+
+                // handle dummy UTF-16 and UTF-32 by scanning for BOM, as in MRI
+                if (encoding == Encodings.UTF16_DUMMY && byteArray.getLength() >= 2) {
+                    int c0 = byteArray.get(0) & 0xff;
```

Comment:
more of a note: ReadByteNode is much better here if the string happens to be native, as that avoids a copy of the whole string to managed just for reading a couple bytes. I changed to ReadByteNode